File: assets/js/customer/customer.js
const caps = (s) => s.split(" ").map(w => w.charAt(0).toUpperCase() + w.slice(1)).join(" ")

class Customer {
    constructor(info) {
        if (info instanceof FormData) {
            this.data = freeze(this.normalize(info))
        } else {
            this.data = freeze(this.convert(info))
        }

        this.fulldata = null
    }

    normalize(formdata) {
        const temp     = {}
        const norm     = {}

        this.fulldata = array(formdata.entries())

        this.fulldata.forEach(e => { temp[e[0]] = e[1] })

        norm.name = {
            last:   caps(temp['name-last']),
            first:  caps(temp['name-first']),
            middle: caps(temp['name-middle'])
        }

        norm.birth = {
            date:    temp['birth-date'],
            place:   caps(temp['birth-place']),
            country: caps(temp['birth-country'])
        }

        norm.mother = {
            last:   caps(temp['mother-last']),
            first:  caps(temp['mother-first']),
            middle: caps(temp['mother-middle'])
        }

        norm.per_add = {
            no:       temp['home-address-no'],
            street:   temp['home-address-street'],
            barangay: temp['home-address-barangay'],
            city:     temp['home-address-city'],
            province: temp['home-address-province'],
            zip:      temp['home-address-zip']
        }

        norm.pre_add = {
            no:       temp['present-address-no'],
            street:   temp['present-address-street'],
            barangay: temp['present-address-barangay'],
            city:     temp['present-address-city'],
            province: temp['present-address-province'],
            zip:      temp['present-address-zip']
        }

        norm.emp_add = {
            no:       temp['employment-address-no'],
            street:   temp['employment-address-street'],
            barangay: temp['employment-address-barangay'],
            city:     temp['employment-address-city'],
            province: temp['employment-address-province'],
            zip:      temp['employment-address-zip']
        }

        norm.contact = {
            mobile: temp['mobile-no'],
            phone:   temp['home-no'],
            email:  temp['email-address']
        }

        norm.employment = {
            type:     caps(temp['employment-type']),
            gmi:      temp['gmi'],
            industry: caps(temp['employment-industry']),
            name:     caps(temp['employment-name']),
            position: caps(temp['employment-position']),
            contact:  temp['employment-contact']
        }

        norm.gender    = temp['gender']
        norm.nation    = caps(temp['nationality'])
        norm.civil     = temp['civil-status']
        norm.residency = temp['residency']
        norm.mailing   = temp['mailing-address']

        return norm
    }

    convert(info) {
        return {...info}
    }

    get name() {
        const last  = this.data.name.last
        const first = this.data.name.first
        const mid   = this.data.name.middle

        return `${last}, ${first} ${mid}`
    }

    get gender() {
        return this.data.gender
    }

    get birthdate() {
        return this.data.birth.date
    }

    get birthplace() {
        const place   = this.data.birth.place
        const country = this.data.birth.country

        return `${place}, ${country}`
    }

    get nation() {
        return this.data.nation
    }

    get mother() {
        const last  = this.data.mother.last
        const first = this.data.mother.first
        const mid   = this.data.mother.middle

        return `${last}, ${first} ${mid}`
    }

    get homeadd() {
        const n = this.data.per_add.no
        const s = this.data.per_add.street
        const b = this.data.per_add.barangay
        const c = this.data.per_add.city
        const z = this.data.per_add.zip
        const p = this.data.per_add.province

        return `${n}, ${s}, ${b}, ${z} ${c}, ${p}`
    }

    get presadd() {
        const n = this.data.pre_add.no
        const s = this.data.pre_add.street
        const b = this.data.pre_add.barangay
        const c = this.data.pre_add.city
        const z = this.data.pre_add.zip
        const p = this.data.pre_add.province

        return `${n}, ${s}, ${b}, ${z} ${c}, ${p}`
    }

    get mobile() {
        return this.data.contact.mobile
    }

    get phone() {
        return this.data.contact.phone
    }

    get email() {
        return this.data.contact.email
    }

    get emptype() {
        return this.data.employment.type
    }

    get gmi() {
        return this.data.employment.gmi
    }

    get empindustry() {
        return this.data.employment.industry
    }

    get empname() {
        return this.data.employment.name
    }

    get empposition() {
        return this.data.employment.position
    }

    get empcontact() {
        return this.data.employment.contact
    }

    get empadd() {
        const n = this.data.emp_add.no
        const s = this.data.emp_add.street
        const b = this.data.emp_add.barangay
        const c = this.data.emp_add.city
        const z = this.data.emp_add.zip
        const p = this.data.emp_add.province

        return `${n}, ${s}, ${b}, ${z} ${c}, ${p}`
    }

    get residency() {
        return this.data.residency
    }

    get mailing() {
        return this.data.mailing
    }
}

File: src/app/chinabank/model/schemas/Customer.java
package app.chinabank.model.schemas;

import app.chinabank.ChinaBank;
import app.chinabank.model.Address;
import app.chinabank.model.Name;
import app.chinabank.model.exceptions.SchemaCreationException;
import app.chinabank.model.tables.CustomerTable;
import app.chinabank.utils.TimeUtil;
import java.sql.Timestamp;
import java.util.HashMap;
import java.util.LinkedHashMap;
import org.apache.commons.lang.RandomStringUtils;
import org.json.simple.JSONObject;

public class Customer implements Schema {

  public String                  id;
  public Timestamp               dateCreated;
  public String                  branchCreated;
  public String                  gender;
  public String                  nationality;
  public String                  civilStatus;
  public String                  residency;
  public String                  mailingAddress;
  public Name                    name;
  public Name                    motherName;
  public Address                 permanentAddress;
  public Address                 presentAddress;
  public Address                 workAddress;
  public Timestamp               birthDate;
  public String                  birthPlace;
  public String                  birthCountry;
  public HashMap<String, String> contact;
  public HashMap<String, String> employment;

  /**
   * Constructor.
   *
   * @param info - customer information
   * @throws SchemaCreationException when an error occured while creating customer.
   */
  @SuppressWarnings("unchecked")
  public Customer(JSONObject info) throws SchemaCreationException {
    this.id               = this.generateID();
    this.dateCreated      = TimeUtil.currentTimestamp();
    this.branchCreated    = ChinaBank.BRANCH_ID;
    this.gender           = (String) info.get("gender");
    this.nationality      = (String) info.get("nation");
    this.civilStatus      = (String) info.get("civil");
    this.residency        = (String) info.get("residency");
    this.mailingAddress   = (String) info.get("mailing");
    this.name             = new Name((JSONObject) info.get("name"));
    this.motherName       = new Name((JSONObject) info.get("mother"));
    this.permanentAddress = new Address((JSONObject) info.get("per_add"));
    this.presentAddress   = new Address((JSONObject) info.get("pre_add"));
    this.workAddress      = new Address((JSONObject) info.get("emp_add"));

    JSONObject birth = (JSONObject) info.get("birth");

    this.birthDate    = TimeUtil.toTimestamp((String) birth.get("date"));
    this.birthPlace   = (String) birth.get("place");
    this.birthCountry = (String) birth.get("country");

    this.contact    = (HashMap<String, String>) info.get("contact");
    this.employment = (HashMap<String, String>) info.get("employment");
  }

  /**
   * Constructor.
   *
   * @param info - Database record
   * @throws SchemaCreationException when an error occured while creating customer
   */
  public Customer(LinkedHashMap<String, Object> info) throws SchemaCreationException {
    this.name                      = new Name();
    this.motherName                = new Name();
    this.presentAddress            = new Address();
    this.permanentAddress          = new Address();
    this.workAddress               = new Address();
    this.contact                   = new HashMap<>();
    this.employment                = new HashMap<>();
    this.id                        = (String) info.get("CustomerID");
    this.dateCreated               = (Timestamp) info.get("DateCreated");
    this.branchCreated             = (String) info.get("BranchID");
    this.name.last                 = (String) info.get("LName");
    this.name.first                = (String) info.get("FName");
    this.name.middle               = (String) info.get("MName");
    this.motherName.last           = (String) info.get("MotherFName");
    this.motherName.first          = (String) info.get("MotherMName");
    this.motherName.middle         = (String) info.get("MotherLName");
    this.gender                    = (String) info.get("Gender");
    this.nationality               = (String) info.get("Nationality");
    this.civilStatus               = (String) info.get("CivilStatus");
    this.residency                 = (String) info.get("Residency");
    this.mailingAddress            = (String) info.get("MailingAdd");
    this.birthDate                 = (Timestamp) info.get("BirthDate");
    this.birthPlace                = (String) info.get("BirthPlace");
    this.birthCountry              = (String) info.get("BirthCountry");
    this.permanentAddress.unit     = (String) info.get("PerAddUnit");
    this.permanentAddress.street   = (String) info.get("PerAddStreet");
    this.permanentAddress.village  = (String) info.get("PerAddVillage");
    this.permanentAddress.city     = (String) info.get("PerAddCity");
    this.permanentAddress.province = (String) info.get("PerAddProvince");
    this.permanentAddress.zip      = (String) info.get("PerAddZip");
    this.presentAddress.unit       = (String) info.get("PreAddUnit");
    this.presentAddress.street     = (String) info.get("PreAddStreet");
    this.presentAddress.village    = (String) info.get("PreAddVillage");
    this.presentAddress.city       = (String) info.get("PreAddCity");
    this.presentAddress.province   = (String) info.get("PreAddProvince");
    this.presentAddress.zip        = (String) info.get("PreAddZip");
    this.workAddress.unit          = (String) info.get("EmpAddUnit");
    this.workAddress.street        = (String) info.get("EmpAddStreet");
    this.workAddress.village       = (String) info.get("EmpAddVillage");
    this.workAddress.city          = (String) info.get("EmpAddCity");
    this.workAddress.province      = (String) info.get("EmpAddProvince");
    this.workAddress.zip           = (String) info.get("EmpAddZip");
    this.contact.put("mobile", (String) info.get("MobileNo"));
    this.contact.put("phone", (String) info.get("PhoneNo"));
    this.contact.put("email", (String) info.get("Email"));
    this.employment.put("gmi", (String) info.get("GMI"));
    this.employment.put("type", (String) info.get("EmpType"));
    this.employment.put("industry", (String) info.get("EmpIndustry"));
    this.employment.put("name", (String) info.get("EmpName"));
    this.employment.put("position", (String) info.get("EmpPosition"));
    this.employment.put("contact", (String) info.get("EmpContact"));
  }

  private String generateID() throws SchemaCreationException {
    int count = CustomerTable.getInstance().count();

    if (count != -1) {
      int    pad         = 6;
      int    numericSize = 9;
      int    checksum    = 0;
      String id          = "C-";
      String countString = Integer.toString(count);
      String numeric     = RandomStringUtils.randomNumeric(numericSize);

      while (countString.length() != pad) {
        countString = '0' + countString;
      }

      id += countString + "-";
      id += numeric + "-";

      char[] numericArr = numeric.toCharArray();

      for (char num : numericArr) {
        int numInt = Integer.parseInt(String.valueOf(num));

        checksum += numInt;
      }

      checksum %= numericSize;

      id += checksum;

      return id;
    } else {
      throw new SchemaCreationException("error generating customer id");
    }
  }

  public String getIdentifier() {
    return this.id.substring(9, 18);
  }

  public String getChecksum() {
    return "" + this.id.charAt(19);
  }

  @Override
  @SuppressWarnings("unchecked")
  public String toJSON() {
    JSONObject birth = new JSONObject();

    birth.put("date", TimeUtil.dateFormat(this.birthDate));
    birth.put("place", this.birthPlace);
    birth.put("country", this.birthCountry);

    JSONObject json = new JSONObject();

    json.put("id", this.id);
    json.put("date_created", TimeUtil.dateFormat(this.dateCreated));
    json.put("branch_created", this.branchCreated);
    json.put("name", this.name.toJsonObject());
    json.put("birth", birth);
    json.put("mother", this.motherName.toJsonObject());
    json.put("per_add", this.permanentAddress.toJsonObject());
    json.put("pre_add", this.presentAddress.toJsonObject());
    json.put("emp_add", this.workAddress.toJsonObject());
    json.put("contact", new JSONObject(this.contact));
    json.put("employment", new JSONObject(this.employment));
    json.put("gender", this.gender);
    json.put("nation", this.nationality);
    json.put("civil", this.civilStatus);
    json.put("residency", this.residency);
    json.put("mailing", this.mailingAddress);

    return json.toJSONString();
  }

  /**
   * Convert customer to a preview JSON string.
   *
   * @return JSON string
   */
  @SuppressWarnings("unchecked")
  public String toFullJSON() {
    JSONObject birth = new JSONObject();

    birth.put("date", TimeUtil.dateFormat(this.birthDate));
    birth.put("place", this.birthPlace);
    birth.put("country", this.birthCountry);

    JSONObject address = new JSONObject();

    address.put("permanent", this.permanentAddress.toJsonObject());
    address.put("present", this.presentAddress.toJsonObject());
    address.put("employment", this.workAddress.toJsonObject());

    JSONObject contact = new JSONObject();

    contact.put("mobile", this.contact.get("mobile"));
    contact.put("phone", this.contact.get("phone"));
    contact.put("email", this.contact.get("email"));
    contact.put("mailing", this.mailingAddress);

    JSONObject json = new JSONObject();

    json.put("id", this.id);
    json.put("name", this.name.toJsonObject());
    json.put("gender", this.gender);
    json.put("nationality", this.nationality);
    json.put("civilStatus", this.civilStatus);
    json.put("birth", birth);
    json.put("mother", this.motherName.toJsonObject());
    json.put("residency", this.residency);
    json.put("dateCreated", TimeUtil.timestampFormat(this.dateCreated));
    json.put("contact", contact);
    json.put("address", address);
    json.put("employment", employment);

    return json.toJSONString();
  }

}
